### ISSUE TITLE
Register TestExecutionListener specific to a LauncherDiscoveryRequest

### ISSUE THREAD

**Issue by noopur2507** [junit-team/junit5#628]
I am using a single instance of `Launcher` to execute multiple requests. For each request, I want to register a separate instance of `TestExecutionListener` with the launcher so that a listener is notified of the events only for that particular request/test plan. The listener should also provide access to the `TestPlan` to which it is listening.  

I am using the existing APIs in Eclipse in the following way:
- Registering a single static listener with the launcher for all the requests being executed. This is not correct as we don't want to use a single static listener.
- Accessing the test plan in the listener via `TestExecutionListener.testPlanExecutionStarted(TestPlan testPlan)` API. This could be a problem if test plans are being run in parallel.

You can have a look at these files for the current implementation: 
- [JUnit5TestReference.java](https://github.com/noopur2507/eclipse-junit5/blob/master/org.eclipse.jdt.junit5.runtime/src/org/eclipse/jdt/internal/junit5/runner/JUnit5TestReference.java)
- [JUnit5TestListener.java](https://github.com/noopur2507/eclipse-junit5/blob/master/org.eclipse.jdt.junit5.runtime/src/org/eclipse/jdt/internal/junit5/runner/JUnit5TestListener.java)

Please let me know if I am missing something or not using these APIs in the intended way.

**Comment by sbrannen** [comment]
Hi @noopur2507,

You're not missing anything: the architecture was never designed to support registration of  a `TestExecutionListener` per `LauncherDiscoveryRequest`.

Rather, a `TestExecutionListener` is registered for a specific `Launcher` instance.

Is it a problem for you to create multiple `Launcher` instances? If so, then the architecture will have to be changed to support such use cases; however, that will likely introduce breaking changes in the current API.

@junit-team/junit-lambda, anyone else have further input on this?

**Comment by sbrannen** [comment]
@noopur2507, while we have your attention, would you mind commenting on https://github.com/junit-team/junit5/issues/504#issuecomment-269302164? 😉

**Comment by noopur2507** [comment]
> Is it a problem for you to create multiple Launcher instances?

No, creating separate `Launcher` instances is not a problem and solves the issue. But for a single launch having multiple requests, it results in an unnecessary overhead of discovering/loading the test engines with each instance of `Launcher`.

**Comment by noopur2507** [comment]
> @noopur2507, while we have your attention, would you mind commenting on #504 (comment)?

Done.

**Comment by sbrannen** [comment]
> But for a single launch having multiple requests, it results in an unnecessary overhead of discovering/loading the test engines with each instance of `Launcher`.

Yes, that's true. But... out of curiosity, how long does that take on your machine? 

I would imagine that's actually a pretty fast process, though I admittedly have not benchmarked it.

**Comment by noopur2507** [comment]
It is fast i.e. takes around 40 ms for `LauncherFactory.create()` call for the first time. For further invocations in the same run, it reduces to 2-4 ms. So looks like some internal caching/optimization is available.

Another issue with multiple `Launcher` instances could be related to memory consumption.

It can be used as the solution for now but would not be the preferred way to handle it.

**Comment by marcphilipp** [comment]
@noopur2507 Thanks for checking!

We could add an overloaded `execute(LauncherDiscoveryRequest, TestExecutionListener)` method that uses the supplied `TestExecutionListener` in addition to already registered listeners but only for the supplied `LauncherDiscoveryRequest`.

@junit-team/junit-lambda What do you guys think?

**Comment by sbrannen** [comment]
> @noopur2507 Thanks for checking!

👍 

> We could add an overloaded `execute(LauncherDiscoveryRequest, TestExecutionListener)` method that uses the supplied `TestExecutionListener` in addition to already registered listeners but only for the supplied `LauncherDiscoveryRequest`.

Yes. That seems plausible.

**Comment by noopur2507** [comment]
+1.

**Comment by sormuras** [comment]
Addressed by https://github.com/junit-team/junit5/commit/e99a78098dae04b1fd2c5cc28ff2a09ab77567e4 
Next ticket https://github.com/junit-team/junit5/issues/740
